### PR TITLE
feat: move getter methods to Result

### PR DESCRIPTION
**Pull request by redeboer**
Moves the `get_intermediate_state_names` function into `Result`. Also defined related function `Result.get_initial_state` and `Result.get_final_state` and implements them in the `amplitude` module.

### CODE REVIEW

File: expertsystem/reaction/solving.py [hunk 1]
```diff
@@ -146,6 +146,41 @@ def extend(
                 else:
                     self.__violated_rules[key].update(rules2)
 
+    def get_initial_state(self) -> List[Particle]:
+        if len(self.solutions) == 0:
+            raise ValueError(
+                f"No solutions in {self.__class__.__name__} object"
+            )
```
Comment (spflueger) [inline expertsystem/reaction/solving.py:153]: Thats already done in `__get_first_graph()`

File: expertsystem/reaction/solving.py [hunk 2]
```diff
@@ -146,6 +146,41 @@ def extend(
...
+        graph = self.__get_first_graph()
+        return [
+            graph.edge_props[i][0] for i in graph.get_initial_state_edges()
+        ]
+
+    def get_final_state(self) -> List[Particle]:
+        if len(self.solutions) == 0:
+            raise ValueError(
+                f"No solutions in {self.__class__.__name__} object"
+            )
```
Comment (spflueger) [inline expertsystem/reaction/solving.py:163]: Same

**Review by spflueger — COMMENTED** [review]
The whole Result class still feels strange. I will approve the PR but we should think about the design in total. I don't like that Result is passed (with violated rules info) into the amp builders. Currently I'm thinking about slipping up the Result class. Then as a result a simple tuple of the solutions and the violated rules is returned. If a solver cannot execute all rules, then an exception could be raised instead of returning that information in the Result as well. I think that was the original design.... maybe that better then trying to combine everything in this Result class.

File: expertsystem/reaction/solving.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/reaction/solving.py]: Ah thanks, weird mistake

**Comment by redeboer** [comment]
> The whole Result class still feels strange

I agree. It's a bit related to #325 though, that I think it's good to return a 'navigable' and safe class, as in, you can probe its contents through methods and it is ensured that it only contains `StateTransitionGraph` instances with consistent initial and final state. But that's more narrowed down than a `Result` class (too general naming btw) that has the responsibility (?) of keeping track of which rules were violated.

**Comment by redeboer** [comment]
Actually for #274, I convert the `Result` back into an exception, which carries all violated rules as `args` of the exception. So that's already a code smell I guess

**Review by spflueger — APPROVED** [review]

**Pull request merged**